libdrgn/python/drgnpy.h

```diff
@@ -97,13 +97,17 @@ typedef struct {
 	struct pyobjectp_set objects;
 } Program;
 
-typedef struct _GenericIterator {
+typedef struct {
+	PyObject_HEAD
+	Program *prog;
```

Comment:
`struct drgn_thread` already stores the `struct drgn_program`, so rather than storing the redundant `Program *`, you can get it as `container_of(thread->thread.prog, Program, prog)`. See `DrgnObject_prog()`.

The same thing also applies to `ThreadIterator`.

Reply:
Maybe I'm just having a momentary bout of confusion, but I _think_ there's an issue with this:

In order for this to be applicable to `ThreadIterator`, we'd have to expose the definition of `struct drgn_thread_iterator` in `program.h` in order for `container_of` to be able to resolve correctly (otherwise you run into an incomplete type error for `struct drgn_thread_iterator`). Then the issue becomes the fact that `struct drgn_thread_iterator` includes a `struct drgn_thread_set_iterator`, which _itself_ would then result in an incomplete type error, because it is defined by the `DEFINE_HASH_TABLE_FUNCTIONS` macro, and we _don't_ want to pull that into the header per your previous comments. Not sure if there's any way to resolve this without pulling that macro into the header though.

Reply:
You're right, just leave it as is for `ThreadIterator`.